### ISSUE TITLE
Enhancement: disable specific log messages

### ISSUE THREAD

**Issue by edreamleo** [flexxui/flexx#520]
As a workaround to suppressing unwanted messages from ace, leoflexx.leo contains this code:
```python
f __name__ == '__main__':
    flx.launch(LeoApp)
    allowed = r'(Critical|Error|Leo|Session|Starting|Stopping|Warning)'
    pattern = re.compile(allowed, re.IGNORECASE)
    flx.set_log_level('INFO', pattern)
    flx.logger.info('LeoApp: after flx.launch')
    flx.run()
```
Instead, it would be better to "say what I mean":
```python
flx.set_log_level('INFO', ignore="Automatically scrolling cursor")
# or...
flx.set_log_level('INFO', ignore=[
   "Unwanted message 1",
   "Unwanted message 2",
])
```
Presumably, this could be done in _MatchFilter.filter in flexx/util/logging.py.

Aside: My old Sherlock tracing product allowed even more flexibility: a list of patterns, to be applied in order, that either enabled or disabled tracing: `["+x", "-y", "+z"]` where x, y and z could be any pattern.  This list was applied to the function's name, and tracing was enabled or disabled as soon as a pattern matched.  This is surely possible in _MatchFilter (match against record.funcName), but it's probably not worth documenting :-)  Otoh, being able to suppress one or more patterns *is* worth doing, imo.

**Comment by edreamleo** [comment]
Hmm.  What to do if both enabling and disabling patterns are present?  In my use case, disabling should have priority.  But perhaps that's contentious.  If so, maybe the Sherlock way would let the user decide. Something like:
```python
flx.set_log_level('INFO', filters=[
   ("disable", "(possibly regex) pattern 1"),
   ("enable", "Desired (possibly regex) pattern 1"),
   ...
])
```
But this might be low priority.  Certainly in my use case an ignore list suffices.

**Comment by almarklein** [comment]
It's an interesting thought. I've never felt the need for such precise filtering myself, nor do I know how other projects handle this. Would be interesting to add at somepoint, but probably not a hight priority indeed.

**Comment by edreamleo** [comment]
@almarklein I'd like to push back a little bit, and suggest that this item should have a reasonably high priority.

Previously, I muddied the waters by talking about various more complicated ways to disable logs.  Instead, I would be happy with a single new **disable arg** .  As with the enabling arg, this would either be a string or a compiled regex. Obviously, the regex could disable multiple messages using regex alternation:
```python
disable = re.compile(r'(message1|message2)')
```
So _MatchFilter.filter would be something like this (not tested):
```python
def filter(self, record):
    match = self.match
    suppress = self.suppress # The suppression arg.
    if not match and not suppress:
        return True
    # Look at suppressions first.
    if isinstance(suppress, str):
        if (suppress in record.name or
            suppress in record.getMessage() or
            suppress in record.funcName
        ):
            return False
    elif suppress:
        if (re.search(suppress, record.name) or
            re.search(suppress, record.getMessage()) or
            re.search(suppress, record.funcName)
        ):
            return False
    # Previous code
    if isinstance(match, str):
        return (match in record.name or
                match in record.getMessage() or
                match in record.funcName)
    else:
        return (re.search(match, record.name) or
                re.search(match, record.getMessage()) or
                re.search(match, record.funcName))
```
This doesn't seem too difficult ;-)  However, setting self.suppress would take a little more doing.

**Comment by edreamleo** [comment]
Here is why this issue is important to me.  Without the proposed enhancement, I can't safely enable any set of messages, because then there is a real danger that I won't enable some error message.

I got bit by this yesterday: no messages, but unexpected results.  It took awhile to remember to turn the debug switch back on so all message would be visible. I then saw the tracebacks. In other words, this code doesn't enable enough:
```python
allowed = r'(Critical|Error|Leo|Session|Starting|Stopping|Warning)'
pattern = re.compile(allowed, re.IGNORECASE)
flx.set_log_level('INFO', pattern)
````
Sure, I could try to add some more enablers, but there would be no guarantee that these patterns would allow all important errors.  But heh, as I write this I see that adding 'Traceback' to the list would be a good idea ;-)

In short, the only truly safe workaround during testing is to allow all log messages, which is annoying.

**Comment by almarklein** [comment]
I can see your use case, but I'm also curious to what kind of messages are so abundant and annoying that you want to filter them. Before diving into this and making the logger system more complex, maybe there's a simpler solution? E.g. what about simply setting log level to 'WARNING'? Or perhaps these annoying messages should use `logger.debug` instead of `logger.info`?

**Comment by edreamleo** [comment]
On Tue, Nov 20, 2018 at 4:53 AM Almar Klein <notifications@github.com>
wrote:

> I can see your use case, but I'm also curious to what kind of messages are
> so abundant and annoying that you want to filter them. Before diving into
> this and making the logger system more complex, maybe there's a simpler
> solution? E.g. what about simply setting log level to 'WARNING'? Or perhaps
> these annoying messages should use logger.debug instead of logger.info?
>
Why didn't I think of that?  The messages are:

[W 09:45:42 flexx.app] JS: Automatically scrolling cursor...

Setting the level to 'ERROR' works just fine.

Edward

**Comment by edreamleo** [comment]
I'm good with closing this.  Thanks for your help.

**Comment by almarklein** [comment]
That's a log message from your code, right? Maybe consider logging such a message using `logger.info()` or `logger.debug()`; it does not look like a warning to me :)

**Comment by edreamleo** [comment]
On Wed, Nov 21, 2018 at 2:11 AM Almar Klein <notifications@github.com>
wrote:

> That's a log message from your code, right?
It's a message from the ace editor.

> Maybe consider logging such a message using logger.info() or
> logger.debug();
>
> it does not look like a warning to me :)

Huh? Of course it's a warning:

[W 09:45:42 flexx.app] JS: Automatically scrolling cursor...

> Maybe consider logging such a message using logger.info() or
logger.debug();

Both show the unwanted message.

Edward
